spatialdata/_core/models.py

```diff
@@ -658,38 +667,35 @@ def parse(
             region_key = attr[cls.REGION_KEY_KEY]
             instance_key = attr[cls.INSTANCE_KEY]
 
-        if isinstance(region, str):
-            if region_key is not None:
-                raise ValueError(
-                    f"If `{cls.REGION_KEY}` is of type `str`, `{cls.REGION_KEY_KEY}` must be `None` as it is redundant."
-                )
-            if instance_key is None:
-                raise ValueError("`instance_key` must be provided if `region` is of type `List`.")
-        elif isinstance(region, list):
-            if region_key is None:
-                raise ValueError(f"`{cls.REGION_KEY_KEY}` must be provided if `{cls.REGION_KEY}` is of type `List`.")
-            if not adata.obs[region_key].isin(region).all():
-                raise ValueError(f"`adata.obs[{region_key}]` values do not match with `{cls.REGION_KEY}` values.")
-            if not is_categorical_dtype(adata.obs[region_key]):
-                warn(f"Converting `{cls.REGION_KEY_KEY}: {region_key}` to categorical dtype.", UserWarning)
-                adata.obs[region_key] = pd.Categorical(adata.obs[region_key])
-            if instance_key is None:
-                raise ValueError("`instance_key` must be provided if `region` is of type `List`.")
-        else:
-            if region is not None:
-                raise ValueError(f"`{cls.REGION_KEY}` must be of type `str` or `List`.")
+        if region_key is None:
+            raise ValueError(f"`{cls.REGION_KEY_KEY}` must be provided.")
+        if isinstance(region, np.ndarray):
+            region = region.tolist()
+        if region is None:
+            raise ValueError(f"`{cls.REGION_KEY}` must be provided.")
+        region_ = region if isinstance(region, list) else [region]
+        if TYPE_CHECKING:
+            assert isinstance(region_, list)
+        if not adata.obs[region_key].isin(region_).all():
+            raise ValueError(f"`adata.obs[{region_key}]` values do not match with `{cls.REGION_KEY}` values.")
+        if not is_categorical_dtype(adata.obs[region_key]):
+            logger.warning(f"Converting `{cls.REGION_KEY_KEY}: {region_key}` to categorical dtype.")
+            adata.obs[region_key] = pd.Categorical(adata.obs[region_key])
+        if instance_key is None:
+            raise ValueError("`instance_key` must be provided.")
 
-        # TODO: check for `instance_key` values?
         attr = {"region": region, "region_key": region_key, "instance_key": instance_key}
         adata.uns[cls.ATTRS_KEY] = attr
         return adata
 
 
 # TODO: consider removing if we settle with geodataframe
-def _sparse_matrix_from_assignment(
+def _sparse_matrix_from_assignment(  # pragma: no cover
     n_obs: int, var_names: Union[list[str], ArrayLike], assignment: pd.Series
 ) -> csr_matrix:
     """Create a sparse matrix from an assignment array."""
+    from tqdm import tqdm
```

Comment:
`tqdm` has been removed from the dependencies of the package. But here I would remove the function `_sparse_matrix_from_assignment()`. It's a legacy from when we where representing points with `AnnData`.